### PR TITLE
perf: Server-Side Data Fetching in App Router: `/availability`

### PR DESCRIPTION
**Pull request by hbjORbj**
## What does this PR do?

- Fixes CAL-4283

## Mandatory Tasks (DO NOT REMOVE)

- [x] I have self-reviewed the code (A decent size PR without self-review might be rejected).
- [x] N/A - I have added a Docs issue [here](https://github.com/calcom/docs/issues/new) if this PR makes changes that would require a [documentation change](https://docs.cal.com/docs). If N/A, write N/A here and check the checkbox.
- [x] I confirm automated tests are in place that prove my fix is effective or that my feature works.

## How should this be tested?

- Navigate to `/future/availability` and `/future/availability/[schedule]`

### CODE REVIEW

File: apps/web/app/future/availability/[schedule]/page.tsx [hunk 1]
```diff
@@ -23,4 +29,36 @@ export const generateMetadata = async ({ params }: PageProps) => {
...
+  const session = await getServerSession(AUTH_OPTIONS);
+  const userId = session?.user?.id ?? -1;
+  const scheduleId = params?.schedule ? Number(params.schedule) : -1;
+  const [schedule, travelSchedules] = await Promise.all([
+    ScheduleRepository.findDetailedScheduleById({
+      scheduleId,
+      isManagedEventType: false,
+      userId,
+      timeZone: "UTC",
+      defaultScheduleId: -1,
```
Comment (hbjORbj) [inline apps/web/app/future/availability/[schedule]/page.tsx:44]: it seems like `session.user` doesn't include `timeZone` or `defaultScheduleId`. Where can I obtain these props? Or should we modify the `AUTH_OPTIONS` logic to acquire these props?

File: apps/web/app/future/availability/[schedule]/page.tsx [hunk 2]
```diff
@@ -23,4 +29,36 @@ export const generateMetadata = async ({ params }: PageProps) => {
...
+
+  return (
+    <PageWrapper
+      getLayout={null}
+      requiresLicense={false}
+      nonce={nonce}
+      themeBasis={null}
+      isBookingPage={false}>
+      <AvailabilitySettingsWebWrapper
+        schedule={schedule as unknown as AvailabilitySettingsProps["schedule"]}
```
Comment (hbjORbj) [inline apps/web/app/future/availability/[schedule]/page.tsx:57]: TODO: fix type

File: apps/web/app/future/availability/[schedule]/page.tsx [hunk 3]
```diff
@@ -23,4 +29,41 @@ export const generateMetadata = async ({ params }: PageProps) => {
...
+    },
+  });
+
+  const scheduleId = params?.schedule ? Number(params.schedule) : -1;
+  const [schedule, travelSchedules] = await Promise.all([
+    ScheduleRepository.findDetailedScheduleById({
+      scheduleId,
+      isManagedEventType: false,
+      userId,
+      timeZone: userData?.timeZone ?? "UTC",
```
Comment (Udit-takkar) [inline apps/web/app/future/availability/[schedule]/page.tsx:51]:
```suggestion
      timeZone: userData?.timeZone ?? "Europe/London",
```

File: packages/lib/server/repository/user.ts [hunk 4]
```diff
@@ -72,6 +72,29 @@ const userSelect = Prisma.validator<Prisma.UserSelect>()({
 });
 
 export class UserRepository {
+  static async findUserByIdWithOptionalSelect({
```
Comment (Udit-takkar) [inline packages/lib/server/repository/user.ts:75]: NIT: I think we can reuse `findById` here instead of creating this new function

**Review by Udit-takkar — COMMENTED** [review]
Code works fine. Left few comments

**Review by zomars — DISMISSED** [review]

File: packages/lib/server/repository/user.ts (hunk 4 above)
Comment (hbjORbj) [reply packages/lib/server/repository/user.ts:75]: addressed

File: apps/web/app/future/availability/[schedule]/page.tsx (hunk 3 above)
Comment (hbjORbj) [reply apps/web/app/future/availability/[schedule]/page.tsx:51]: addressed

**Comment by hbjORbj** [comment]
@zomars Hey Omar, can I get a re-approval? Just added some refactors ;) Sorry about the hassle

File: apps/web/app/future/availability/[schedule]/page.tsx [hunk 5]
```diff
@@ -23,4 +28,39 @@ export const generateMetadata = async ({ params }: PageProps) => {
   );
 };
 
-export default WithLayout({ getLayout: null, Page: AvailabilitySettingsWebWrapper });
+const Page = async ({ params }: PageProps) => {
+  const session = await getServerSession(AUTH_OPTIONS);
+  const userId = session?.user?.id ?? -1;
```
Comment (joeauyeung) [inline apps/web/app/future/availability/[schedule]/page.tsx:33]: If there's no userId found can we early return the `notFound()`?

File: apps/web/app/future/availability/[schedule]/page.tsx [hunk 6]
```diff
@@ -23,4 +28,39 @@ export const generateMetadata = async ({ params }: PageProps) => {
...
+      select: {
+        timeZone: true,
+        defaultScheduleId: true,
+      },
+    });
+  } catch (e) {
+    notFound();
+  }
+
+  const scheduleId = params?.schedule ? Number(params.schedule) : -1;
```
Comment (joeauyeung) [inline apps/web/app/future/availability/[schedule]/page.tsx:48]: Same here with with the early return?

File: apps/web/app/future/availability/page.tsx [hunk 7]
```diff
@@ -10,4 +14,20 @@ export const generateMetadata = async () => {
...
 };
 
-export default WithLayout({ getLayout: null, Page: AvailabilityPage });
+const Page = async () => {
+  const session = await getServerSession(AUTH_OPTIONS);
+  const userId = session?.user?.id ?? -1;
+  const orgId = session?.user?.org?.id ?? -1;
+
+  let currentOrg = null;
+  try {
```
Comment (joeauyeung) [inline apps/web/app/future/availability/page.tsx:23]: I don't think we should we should make the query for the current org if no orgId is present.

File: packages/lib/server/repository/user.ts [hunk 8]
```diff
@@ -238,12 +238,12 @@ export class UserRepository {
     };
   }
 
-  static async findById({ id }: { id: number }) {
+  static async findById({ id, select }: { id: number; select?: Prisma.UserSelect }) {
```
Comment (joeauyeung) [inline packages/lib/server/repository/user.ts:241]: Same comment as here: https://github.com/calcom/cal.com/pull/16449#discussion_r1751155289

**Review by joeauyeung — DISMISSED** [review]
Pages are working but I have some opinions about early returns.

**Review by zomars — APPROVED** [review]

**Pull request merged**